[DON'T SQUASH] Swizzles, part 1 (add in_warp_transpose)

krzysz00 opened the pull request:
Since integrating swizzles would require some reworking of the ThreadwiseCopyV2 lowering, this PR just adds the in_warp_transpose op and a standalone test that confirms that it works correctly, so that when we integrate it into the codebase we have fewer places to poke at for bugs.

This PR contains a mix of internal and external changes which **should not be squashed together**.

krzysz00 commented:
Having looked further and read more carefully, it appears that the problem here is that threadwiseCopyV2 doesn't support legacyLoad/legacyStore, so that's part of the problem

But also, if I call
```
./build/bin/mlir-miopen-driver -fil_layout=kyxc -in_layout=nhwc -out_layout=nhwk -batchsize=256 -in_channels=128 -in_h=28 -in_w=28 -out_channels=128 -fil_h=3 -fil_w=3 --dilation_h=1 --dilation_w=1 --conv_stride_h=1 --conv_stride_w=1 --padding_h=1 --padding_w=1 -t f32 --operation conv2d -x2 -c | less 
```

the buffer stores are like (this example is chosen from the end)
```
gpu.raw_buffer_store(%2366, %arg2, %c0_i32, %2388, %2383, %2378, %24, %2368) : f32, memref<256x28x28x1x128xf32>, i32
```

which indicates that something's up with how the types on all this are being chosen

krzysz00 commented:
Should I un-`skip-ci` this, or do folks prefer to run tests locally?

krzysz00 commented:
Further observations on `ThreadwiseCopyV2Lowering` - it makes no use of the source read length and destination write length.

Random question - @whchung, what's the difference between ThreadwiseCopy and ThreadwiseCopyV2?

whchung commented:
`threadwise_copy` : the source is a memref.
`threadwise_copy_v2` : the source is a vector. While implementing XDLOP lowering path, it was evident that we need a version of `threadwise_copy` which takes vector as input, hence this `_v2` op.

The implementation of the two ops is actually largely the same. I want to do another refactoring to combine them together, after getting the KPACK implementation done.

krzysz00 commented:
Ok, so it looks like there's a lot of missing logic in threadwise_copy_v2 that I'll need to add to make swizzles work...

krzysz00 commented:
TODO: Add a standalone test for the in warp transpose lowering.

krzysz00 commented:
@whchung Can you help me figure out why lowering in broken on a minimal example? (This is what I cut down from trying to run my standalone test).

To run this, use 
```
./bin/mlir-rocm-runner --shared-libs=./lib/librocm-runtime-wrappers.so,./external/llvm-project/llvm/lib/libmlir_runner_utils.so --entry-point-result=i32 minimize.mlir
```
from the build directory or some similar equivalent invocation

[minimize.mlir.txt](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/files/7086021/minimize.mlir.txt)

whchung commented:
> @whchung Can you help me figure out why lowering in broken on a minimal example? (This is what I cut down from trying to run my standalone test).
> 
> To run this, use
> 
> ```
> ./bin/mlir-rocm-runner --shared-libs=./lib/librocm-runtime-wrappers.so,./external/llvm-project/llvm/lib/libmlir_runner_utils.so --entry-point-result=i32 minimize.mlir
> ```
> 
> from the build directory or some similar equivalent invocation
> 
> [minimize.mlir.txt](https://github.com/ROCmSoftwarePlatform/llvm-project-mlir/files/7086021/minimize.mlir.txt)

I don't see any GPU modules inside the file. In this case, could you try run with `mlir-cpu-runner`?

krzysz00 commented:
> I don't see any GPU modules inside the file. In this case, could you try run with `mlir-cpu-runner`?

The result is the rather confusing
```
kdrewnia@ixt-rack-112 ~/llvm-project-mlir/build 19:01 % ./external/llvm-project/llvm/bin/mlir-cpu-runner -e main --entry-point-result=i32 ../minimize.mlir
Error: entry point not found
```
even if I remove the surrounding `module{}`

whchung commented:
> > I don't see any GPU modules inside the file. In this case, could you try run with `mlir-cpu-runner`?
> 
> The result is the rather confusing
> 
> ```
> kdrewnia@ixt-rack-112 ~/llvm-project-mlir/build 19:01 % ./external/llvm-project/llvm/bin/mlir-cpu-runner -e main --entry-point-result=i32 ../minimize.mlir
> Error: entry point not found
> ```
> 
> even if I remove the surrounding `module{}`

Please check tests related with `mlir-cpu-runner`, you are operating at the incorrect dialect. `mlir-cpu-runner` only takes MLIR LLVM dialect. Try add `mlir-opt -convert-std-to-llvm` in your lit `RUN` command.

krzysz00 commented:
@whchung 
```
kdrewnia@ixt-rack-112 ~/llvm-project-mlir/build 18:55 % ./bin/mlir-opt --convert-std-to-llvm ../minimize.mlir
../minimize.mlir:15:10: error: 'std.load' op operand #1 must be index, but got 'i64'
    %5 = load %0[%3] : memref<1024xi32>
         ^
../minimize.mlir:15:10: note: see current operation: %21 = "std.load"(%4, %19) : (memref<1024xi32>, i64) -> i32
```

which is a somewhat more helpful version of the error I'm finding in the first place

(and the test itself does have a bunch of GPU code - minimize.mlir is a cut down version that just shows the broken bit)

whchung commented:
@krzysz00 seems to be something related with `%0 = tensor_to_memref %cst : memref<1024xi32>`. Rewrite that line to something like:
`%0 = alloc() : memref<1024xi32>` gets the file lowered.

krzysz00 commented:
@whchung From the looks of it, I was missing the magic incantations `--tensor-constant-bufferize --finalizing-bufferize` in the mlir-opt call

krzysz00 committed:
Use select and not if in in_warp_transpose lowering

This workaround, suggested at
https://ontrack-internal.amd.com/browse/SWDEV-302609
allows the code that implements InWarpTranspose to compile correctly.

However, it may lead to increased register pressure when we integrate
swizzles into the broader environment.

krzysz00 committed:
[MLIR][GPU] Allow buffer load/store ops to work on 32-bit integers.

Since these ops work on 32-bit floats, allowing integers should not
affect anything.

krzysz00 committed:
[MLIR][GPU] Add gpu.warp_swizzle and its lowering on AMD GPUs.

The gpu.warp_swizzle is a lowest-common-denominator op for moving data
between threads/lanes on a GPU. It allows arbitrary permutations of
data within groups of 4 adjacent threads, since more complex
permutations may be expensive or only available on some platforms.

(The op that takes a computed lane ID, which also would be supported
everywhere, is not one that is added here as it would perform poorly,
likely require synchronization, and is not needed for the applications
that led to devising this intrinsic.)

This commit also adds an intrinsic to the ROCDL dialect that is
equivalent to a mov with dpp specified, which is used to lower
gpu.warp_swizzle on AMD devices.

krzysz00 committed:
Add miopen.in_warp_transpose, its lowering, and a standalone test

This creates an op, in_warp_transpose that takes a vector of N
values, which are spread over T threads, and transposes each SxS block
within the NxT matrix logically formed by those values, where S is 2
or 4 (for now), optionally permuting the result rows (which thread
gets what) of the transpose. This operation will be used to enable
more efficient stores (and, eventually, loads) in the convolutions we
generate.

krzysz00 commented:
@whchung , now that I have a workaround from the compiler team for the optimizer bug, can you review this (or point it at someone else who has the time?)

whchung commented on external/llvm-project/mlir/include/mlir/Dialect/GPU/GPUOps.td:
```diff
@@ -1141,7 +1141,7 @@ def GPU_MubufLoadOp :
     Arguments<(ins AnyMemRef:$memref,
                    Variadic<I32>:$indices)>,
-    Results<(outs AnyTypeOf<[I16, F16, F32,
+    Results<(outs AnyTypeOf<[I16, F16, F32, I32,
```
could you help say the rationale for adding `I32`?
these relevant intrinsics don't really take LLVM i32 to my understanding.

krzysz00 replied:
This was so that we could load integer values from buffers, instead of floats

whchung replied:
Would casting from I32 -> F32 or I32 -> F16*2 be implemented in GPU->ROCDL lowering?

krzysz00 replied:
No, the point is that we should be able to `buffer_load` an i32 value.

However, having reviewed the test script, we use `memref.load` there, so we don't need to have this change. I can remove it.

whchung commented on external/llvm-project/mlir/lib/Conversion/GPUToROCDL/LowerGpuOpsToROCDLOps.cpp:
```diff
@@ -916,6 +919,46 @@ struct BFOpLowering : ConvertToLLVMPattern {
+        loc, llvmI32Type, rewriter.getI32IntegerAttr(permConst));
+
+    auto noMaskConstImm = rewriter.create<LLVM::ConstantOp>(
+        loc, llvmI32Type, rewriter.getI32IntegerAttr(0b1111));
```
this magic constant value probably deserves a comment.

krzysz00 replied:
Fixed

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -6748,19 +6756,44 @@ struct GridwiseGemmV2RewritePattern
+    auto *ctx = op.getContext();
+
+    const auto &tailResults = blockwiseGemmV2TailOp->getResults();
+#if 0 // ThreadwiseCopyV2 can't handle swizzled results, don't generate them yet
```
can we remove the #if 0 block?

krzysz00 replied:
Probably, yeah. We'll end up recreating it anyway in the followup.

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -7553,6 +7577,310 @@ struct FillRewritePattern : public OpRewritePattern<miopen::FillOp> {
+                                             shouldParticipateVal, zeroConst);
+      }
+
+#if 0 // Compiler is busted, explicitly emit selects
```
can we also remove this `#if 0` block?

krzysz00 replied:
That one I'd like to keep in, since it's got the code I'd _like_ to use once the compiler team fixes the backend.

whchung replied:
Perhaps revise the comment adding TODO?

krzysz00 replied:
Done

whchung commented on mlir/lib/Dialect/MIOpen/Transforms/LowerMIOpenOps.cpp:
```diff
@@ -152,6 +152,7 @@ void LowerMIOpenOpsStep3Pass::runOnOperation() {
 void LowerMIOpenOpsStep4Pass::runOnOperation() {
   MLIRContext *ctx = &getContext();
   OwningRewritePatternList patterns(ctx);
+    patterns.insert<InWarpTransposeRewritePattern>(ctx);
```
nit: clang-format would likely fail on this line?

krzysz00 replied:
Fixed

whchung commented on mlir/lib/ExecutionEngine/ROCm/rocm-runtime-wrappers.cpp:
```diff
@@ -279,6 +279,36 @@ extern "C" void mgpuMemCopy(float *sourceAllocated, float *sourceAligned,
             static_cast<hipMemcpyKind>(copyDirection));
 }
 
+extern "C" StridedMemRefType<int32_t, 1>
```
are these library routines added because in_warp_transpose takes i32?

krzysz00 replied:
Yep

krzysz00 replied:
(Well, mainly so I can have an integer test instead of a float one)

whchung reviewed (commented):
Did the first iteration. Will do another iteration on more involved logic.

whchung commented:
Adding `skip-ci` label to temporarily skipping CI for this PR. Will re-enable it after #396 gets merged.

whchung commented on mlir/include/mlir/Dialect/MIOpen/MIOpenOps.td:
```diff
@@ -490,4 +483,51 @@ def MIOpen_DataConvertOp :
   }];
 }
+
+// TODO: upstream these
```
nit picking: perhaps add your ID here so it's easier to grep for TODO items.

krzysz00 replied:
Fixed

whchung commented on mlir/lib/Dialect/MIOpen/MIOpenOps.cpp:
```diff
@@ -173,6 +177,53 @@ static LogicalResult verify(ThreadwiseCopyOp op) {
+// InWarpTransposeOp
+//===----------------------------------------------------------------------===//
+
+static constexpr size_t warpGroupSize = 4;
```
nit: I guess this `4` is the same value defined for `ins VectorOfDivisibleLengthAndType<4, [F32, I32]>:$vector` in `MIOpen_InWarpTransposeOp`?

In such case how about we define this constant in `extraClassDeclaration`?

krzysz00 replied:
I think I did what you were looking for

whchung commented on mlir/test/Dialect/MIOpen/integration/in_warp_transpose/inWarpTransposeTest.py:
```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
+# RUN: %python %s | miopen-opt \
```
After the surprises in PR #385, zombie processes, and our offline discussions, wondering can we change this test so the script is executed offline, generate test patterns offline?

krzysz00 replied:
Yeah, that's a perfectly reasonable change, I can do that.

krzysz00 replied:
Done

whchung reviewed (dismissed):
I'd like to request change to the test script so test patterns are generated offline instead of online.

krzysz00 committed:
Generate inWarpTranspose test code offline

krzysz00 committed:
Reviwe comments on the op

krzysz00 committed:
[fold to external] Revert adding i32 to load/store ops, it isn't needed

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -7841,6 +7850,310 @@ struct FillRewritePattern : public OpRewritePattern<miopen::FillOp> {
+    Right
+  };
+
+  Value emitRotations(Location loc, PatternRewriter &b, Value vector,
```
The logic in this function is non-trivial, perhaps a high-level description would be good?

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -7841,6 +7850,310 @@ struct FillRewritePattern : public OpRewritePattern<miopen::FillOp> {
+                     uint32_t groupSize, uint32_t totalSize,
+                     ArrayRef<uint32_t> inGroupPerm) const {
+
+    llvm::SmallVector<ArrayAttr, 4> swizzlePerms;
```
Could all these `4` be switched to use `swizzleGroupSize`?

whchung commented on mlir/include/mlir/Dialect/MIOpen/LowerMIOpenOps.h:
```diff
@@ -7841,6 +7850,310 @@ struct FillRewritePattern : public OpRewritePattern<miopen::FillOp> {
+    uint32_t groupSize = op.size();
+
+    ArrayAttr inGroupPermAttr = op.inGroupPerm();
+    llvm::SmallVector<uint32_t, 4> inGroupPerm;
```
Same here. Could all these `4` be switched to use `swizzleGroupSize`?

whchung commented on mlir/utils/generateInWarpTransposeTest.py:
```diff
@@ -0,0 +1,205 @@
+
+
+def testHeader() -> str:
+    ret = """// ### DO NOT EDIT ###
```
nit: how about add a newline after `"""`, so the multi-line strings look closer to the output?

krzysz00 committed:
Remove magic constants in inWarpTranspose lowering

whchung reviewed (approved):
LGTM
